library/Backends/llvm/src/Juvix/Backends/LLVM/Pass/ClosureConversion.hs

```diff
@@ -0,0 +1,170 @@
+          ErasedAnn.CatCoproductIntroLeftM p1 ->
+            Types.CatCoproductIntroLeftM (convert p1 oldToNew)
+          ErasedAnn.CatCoproductIntroRightM p1 ->
+            Types.CatCoproductIntroRightM (convert p1 oldToNew)
+          ErasedAnn.CatCoproductElimM t1 t2 t3 t4 t5 ->
+            Types.CatCoproductElimM
+              (convert t1 oldToNew)
+              (convert t2 oldToNew)
+              (convert t3 oldToNew)
+              (convert t4 oldToNew)
+              (convert t5 oldToNew)
+          ErasedAnn.UnitM -> Types.UnitM
+          ErasedAnn.AppM f xs ->
+            Types.AppM (convert f oldToNew) (fmap (flip convert oldToNew) xs)
+   in Types.Ann {usage, annTy = type', term = newTerm}
+
+handleLambda :: (Term, Type') -> Environment -> Types.TermClosure
+handleLambda (ErasedAnn.LamM {arguments, capture, body}, ty) old@Env {ofMap, tyMap}
+  | null capture =
+    Types.LamM {arguments, body = convert body newEnv}
+  | otherwise =
+    let closureArgs = lamArgumentToClosureArgs arguments
+        closureCapt = lamMCapturesToCapture capture old
+        newEnvMapping =
+          foldMap argumentToEnv closureArgs
+            <> foldMap captureToEnv closureCapt
+            <> ofMap -- last as it gets overwritten by the other mapEnvs
+     in Types.Closure
+          { capture = closureCapt,
+            argumentOffsets = closureArgs,
+            body = convert body (newEnv {ofMap = newEnvMapping})
+          }
+  where
+    -- Need to propagate bindings in lambda
+    newEnv = Env {tyMap = typeMapFromPi arguments ty <> tyMap, ofMap}
+handleLambda _ _ =
+  panic "impossible"
+
+handleName :: EnvOffsetMap -> NameSymbol.T -> Types.TermClosure
+handleName ofMap name =
+  case HashMap.lookup name ofMap of
+    Just index -> Types.ArrayIndex index
+    Nothing -> Types.Var name
+
+-----------------------------------
+-- Key Term Pass Functions
+-----------------------------------
+
+nameToCaptured :: NameSymbol.T -> EnvOffsetMap -> Types.CaptureFrom
+nameToCaptured name envMapping =
+  case HashMap.lookup name envMapping of
+    Nothing -> Types.FromAmbientEnv name
+    Just id -> Types.FromClosureEnv id
+
+lamMCapturesToCapture :: [NameSymbol.T] -> Environment -> [Types.Capture]
+lamMCapturesToCapture names Env {ofMap = envMapping, tyMap} =
+  zipWith f names (enumFrom (0 :: Natural))
+  where
+    f name index =
+      Types.Capture
+        { location = nameToCaptured name envMapping,
+          slot = Types.Slot {previousName = name, newIndex = Types.Index index},
+          capType =
+            case HashMap.lookup name tyMap of
+              Just ty -> ty
+              Nothing -> panic "unknown captured argument"
+        }
+
+lamArgumentToClosureArgs :: [NameSymbol.T] -> [Types.ArraySlot]
+lamArgumentToClosureArgs names = zipWith f names (enumFrom (0 :: Natural))
+  where
+    f name idx =
+      Types.Slot {previousName = name, newIndex = Types.Index idx}
```

Comment:
Would it be clearer to factor out the commonality between this and `lamMCapturesToCapture`?  (I realize that the output types are different, so maybe it would just end up looking confusing or obscure.)